# Loading Data

In [ ]:
# Import libraries
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

In [ ]:
# Load the dataset
dataset = pd.read_csv("../../Data/Churn Prediction Data/churn-bigml-20.csv")

print(dataset.head())

  State  Account length  Area code International plan Voice mail plan  \
0    LA             117        408                 No              No   
1    IN              65        415                 No              No   
2    NY             161        415                 No              No   
3    SC             111        415                 No              No   
4    HI              49        510                 No              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                      0              184.5               97   
1                      0              129.1              137   
2                      0              332.9               67   
3                      0              110.4              103   
4                      0              119.3              117   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             31.37              351.6               80             29.89   
1             21.95   

# Looking for dataset information

In [160]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 667 entries, 0 to 666
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   667 non-null    object 
 1   Account length          667 non-null    int64  
 2   Area code               667 non-null    int64  
 3   International plan      667 non-null    object 
 4   Voice mail plan         667 non-null    object 
 5   Number vmail messages   667 non-null    int64  
 6   Total day minutes       667 non-null    float64
 7   Total day calls         667 non-null    int64  
 8   Total day charge        667 non-null    float64
 9   Total eve minutes       667 non-null    float64
 10  Total eve calls         667 non-null    int64  
 11  Total eve charge        667 non-null    float64
 12  Total night minutes     667 non-null    float64
 13  Total night calls       667 non-null    int64  
 14  Total night charge      667 non-null    fl

Interpretation of the information of the above info data
- `<class 'pandas.core.frame.DataFrame'>` - shows the type of the dataset
- Rows - 667
- Cols - 20
- Shape - `(667,20)`
- How many cols of each type?
    - `bool(1), float64(8), int64(8), object(3)`
    - This helps in identification of **Categorical** cols.
    - Categorical cols - `State`, `International plan`,`Voice mail plan` and `Churn`.

# Handling missing and duplicated values

In [161]:
print(dataset.isna().sum())

State                     0
Account length            0
Area code                 0
International plan        0
Voice mail plan           0
Number vmail messages     0
Total day minutes         0
Total day calls           0
Total day charge          0
Total eve minutes         0
Total eve calls           0
Total eve charge          0
Total night minutes       0
Total night calls         0
Total night charge        0
Total intl minutes        0
Total intl calls          0
Total intl charge         0
Customer service calls    0
Churn                     0
dtype: int64


In [162]:
print(dataset.duplicated().sum())

0


Zero missing and duplicated values. Good to go!

# Categorical Cols

Categorical cols - `State`, `International plan`,`Voice mail plan` and `Churn`.

Columns like `International plan` and `Voice mail plan` has `Yes` and `No` as values which is object type which gets converted to `1` for `Yes` and `0` for `No`.

In [163]:
dataset['International plan'] = dataset['International plan'].map({'Yes': 1, 'No': 0})
dataset['Voice mail plan'] = dataset['Voice mail plan'].map({'Yes': 1, 'No': 0})
print(dataset[['International plan','Voice mail plan']].head())

   International plan  Voice mail plan
0                   0                0
1                   0                0
2                   0                0
3                   0                0
4                   0                0


`Churn` column is boolean which can be replaced by 0 and 1.

In [164]:
dataset['Churn'] = dataset['Churn'].astype(int)
print(dataset[['Churn']].head())

   Churn
0      0
1      1
2      1
3      0
4      0


`State` column is unable to represent in order and has multiple entries which leads to `one-hot encoding`.

In [165]:
dataset = pd.get_dummies(dataset, columns=['State'], drop_first=True)

state_encoded = []
for i in dataset.columns:
    if i.startswith('State'):
        state_encoded.append(i)

for i in state_encoded:
    a = f"{i}"
    dataset[a] = dataset[a].astype(int)
print(dataset.head())

   Account length  Area code  International plan  Voice mail plan  \
0             117        408                   0                0   
1              65        415                   0                0   
2             161        415                   0                0   
3             111        415                   0                0   
4              49        510                   0                0   

   Number vmail messages  Total day minutes  Total day calls  \
0                      0              184.5               97   
1                      0              129.1              137   
2                      0              332.9               67   
3                      0              110.4              103   
4                      0              119.3              117   

   Total day charge  Total eve minutes  Total eve calls  ...  State_SD  \
0             31.37              351.6               80  ...         0   
1             21.95              228.5              

# Features Selection

Rearranging columns to seperate features easily

In [166]:
cols = ["Churn"]
dataset = dataset[cols + [col for col in dataset.columns if col not in cols]]
print(dataset.head())

   Churn  Account length  Area code  International plan  Voice mail plan  \
0      0             117        408                   0                0   
1      1              65        415                   0                0   
2      1             161        415                   0                0   
3      0             111        415                   0                0   
4      0              49        510                   0                0   

   Number vmail messages  Total day minutes  Total day calls  \
0                      0              184.5               97   
1                      0              129.1              137   
2                      0              332.9               67   
3                      0              110.4              103   
4                      0              119.3              117   

   Total day charge  Total eve minutes  ...  State_SD  State_TN  State_TX  \
0             31.37              351.6  ...         0         0         0   
1   

In [167]:
data = dataset.values

In [168]:
X = data[:,1:]
y = data[:, 0]

In [169]:
print("Features = ", X[:5])
print()
print("Target = ", y[:5])

Features =  [[117.   408.     0.     0.     0.   184.5   97.    31.37 351.6   80.
   29.89 215.8   90.     9.71   8.7    4.     2.35   1.     0.     0.
    0.     0.     0.     0.     0.     0.     0.     0.     0.     0.
    0.     0.     0.     0.     0.     1.     0.     0.     0.     0.
    0.     0.     0.     0.     0.     0.     0.     0.     0.     0.
    0.     0.     0.     0.     0.     0.     0.     0.     0.     0.
    0.     0.     0.     0.     0.     0.     0.     0.  ]
 [ 65.   415.     0.     0.     0.   129.1  137.    21.95 228.5   83.
   19.42 208.8  111.     9.4   12.7    6.     3.43   4.     0.     0.
    0.     0.     0.     0.     0.     0.     0.     0.     0.     0.
    0.     0.     1.     0.     0.     0.     0.     0.     0.     0.
    0.     0.     0.     0.     0.     0.     0.     0.     0.     0.
    0.     0.     0.     0.     0.     0.     0.     0.     0.     0.
    0.     0.     0.     0.     0.     0.     0.     0.  ]
 [161.   415.     0.     0.   

Spliting of Dataset - Train/Test set

In [170]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

Feature scaling

In [ ]:
scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [172]:
print("X_train shape = ", X_train.shape)
print("X_test shape = ", X_test.shape)
print("y_train shape = ", y_train.shape)
print("y_test shape = ", y_test.shape)

X_train shape =  (533, 68)
X_test shape =  (134, 68)
y_train shape =  (533,)
y_test shape =  (134,)
